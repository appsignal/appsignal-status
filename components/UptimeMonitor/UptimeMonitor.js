import React from "react";
import PropTypes from "prop-types";

import OutagesOverlay from "../OutagesOverlay";
import UptimeDots from "../UptimeDots";

export const LoadingDot = () => {
  return (
    <div
      data-testid="loadingDot"
      className={`h-8 flex-grow animate-pulse bg-gray-200 rounded`}
    />
  );
};

export const UptimeMonitorLoading = () => {
  const amountOfDots = 30;
  return (
    <div className="flex space-x-1">
      {[...Array(amountOfDots)].map((_, dotNumber) => {
        return <LoadingDot key={dotNumber} />;
      })}
    </div>
  );
};

const UptimeMonitor = ({ uptimeMonitor, threshold }) => {
  const [overlayOpen, setOverlayOpen] = React.useState(false);
  const [loading, setLoading] = React.useState(true);
  const [monitor, setMonitor] = React.useState([]);

  React.useEffect(() => {
    // We use the mounted variable to make sure React stops trying to use state
    // when the component is unmounted.
    let mounted = true;
    fetch(uptimeMonitor.endpoint, { mode: "cors" })
      .then((res) => res.json())
      .then((result) => {
        if (mounted) {
          setMonitor(result);
          setLoading(false);
        }
      });

    return () => (mounted = false);
  }, [uptimeMonitor.id, uptimeMonitor.endpoint]);

  return (
    <>
      <div className="px-6 py-5 space-y-3">
        <div className="sm:flex justify-between">
          <h2>
            <button
              className="c_h-heading focus:outline-none"
              onClick={() => setOverlayOpen(true)}
            >
              {uptimeMonitor.title}
            </button>
          </h2>
          <p className="mt-1 sm:mt-0 text-gray-700">
            Monitoring from {uptimeMonitor.regions.length}{" "}
            {uptimeMonitor.regions?.length > 1 ? "locations" : "location"}
          </p>
        </div>
        {loading && <UptimeMonitorLoading />}
        {!loading && (
          <UptimeDots
            timeseries={monitor.timeseries}
            regions={uptimeMonitor.regions}
            threshold={threshold}
          />
        )}
      </div>

      {!loading && (
        <OutagesOverlay
          open={overlayOpen}
          regions={uptimeMonitor.regions}
          handleClose={() => setOverlayOpen(false)}
          timeseries={monitor.timeseries}
          title={`All outages for ${uptimeMonitor.title}`}
          threshold={threshold}
        />
      )}
    </>
  );
};

UptimeMonitor.propTypes = {
  uptimeMonitor: PropTypes.object.isRequired,
  threshold: PropTypes.number,
};

export default UptimeMonitor;
