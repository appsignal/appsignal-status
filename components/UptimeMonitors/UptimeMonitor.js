import PropTypes from "prop-types";
import Tippy from "@tippyjs/react";

const Bar = () => {
  return (
    <Tippy
      content={
        <div className="text-center text-sm">
          <p className="text-gray-200">July 19th</p>
          <p>No outage</p>
        </div>
      }
    >
      <div className="h-8 flex-grow bg-green-500 rounded" />
    </Tippy>
  );
};

const UptimeMonitor = ({ uptimeMonitor }) => {
  return (
    <div className="px-6 py-5 space-y-3">
      <div className="sm:flex justify-between">
        <h2 className="c_h-heading">{uptimeMonitor.title}</h2>
        <p className="mt-1 sm:mt-0 text-gray-700">
          Monitoring from 4 locations
        </p>
      </div>
      <div className="flex space-x-1">
        {[...Array(30)].map((e, index) => (
          <Bar key={index} />
        ))}
      </div>
    </div>
  );
};

UptimeMonitor.propTypes = {
  uptimeMonitor: PropTypes.object.isRequired,
};

export default UptimeMonitor;
