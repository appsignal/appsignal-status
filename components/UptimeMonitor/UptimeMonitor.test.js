import { render, screen, waitFor } from "@testing-library/react";

import UptimeMonitor, { UptimeMonitorLoading } from "./UptimeMonitor";
import statusPageMock from "../../mocks/status_pages/appsignal.json";

const build = () => {
  return render(
    <UptimeMonitor
      hostname="example.com"
      uptimeMonitor={statusPageMock.uptime_monitors[0]}
    />
  );
};

describe("UptimeMonitor", () => {
  test("renders without errors", () => {
    const { container } = build();
    expect(container).toMatchSnapshot();
  });

  test("renders basic monitor info before loading data", () => {
    build();

    expect(screen.getByText("homepage")).toBeInTheDocument();
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });

  test("renders uptime dots when loading is done", async () => {
    build();

    await waitFor(() => {
      expect(screen.getByText("homepage")).toBeInTheDocument();
      expect(screen.queryByText("Loading...")).not.toBeInTheDocument();
      const uptimeDots = screen.getAllByTestId("uptimeDot");
      expect(uptimeDots.length).toEqual(30);
    });
  });
});

describe("UptimeMonitorLoading", () => {
  test("renders a loading message", () => {
    render(<UptimeMonitorLoading />);
    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });
});
