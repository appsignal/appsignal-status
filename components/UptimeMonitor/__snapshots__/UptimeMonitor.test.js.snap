// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UptimeMonitor renders without errors 1`] = `
<div>
  <div
    class="px-6 py-5 space-y-3"
  >
    <div
      class="sm:flex justify-between"
    >
      <h2>
        <button
          class="c_h-heading focus:outline-none"
        >
          homepage
        </button>
      </h2>
      <p
        class="mt-1 sm:mt-0 text-gray-700"
      >
        Monitoring from 4 locations
      </p>
    </div>
    <div
      class="h-8 text-center"
    >
      Loading...
    </div>
  </div>
</div>
`;
